openlibrary/core/imports.py

```diff
@@ -32,18 +33,56 @@ def load_items(self, filename):
         items = [line.strip() for line in open(filename) if line.strip()]
         self.add_items(items)
 
-    def add_items(self, items):
+    def dedupe_ia_ids(self, items):
+        already_present = [
+            row.ia_id for row in db.query(
+                "SELECT ia_id FROM import_item WHERE ia_id IN $items",
+                vars=locals()
+            )
+        ]
+        # ignore already present
+        logger.info(
+            "batch %s: %d items are already present, ignoring...",
+            self.name,
+            len(already_present)
+        )
+        return list(set(items) - set(already_present))
+
+    def add_items(self, items, ia_items=True):
+        """
+        :param ia_items: Whether items is a list of IA identifiers. If False, expects a list of
+        book data dicts.
+        """
         if not items:
             return
+
         logger.info("batch %s: adding %d items", self.name, len(items))
-        already_present = [row.ia_id for row in db.query("SELECT ia_id FROM import_item WHERE ia_id IN $items", vars=locals())]
-        # ignore already present
-        items = list(set(items) - set(already_present))
 
-        logger.info("batch %s: %d items are already present, ignoring...", self.name, len(already_present))
+        if ia_items:
+            items = self.dedupe_ia_items(items)
+
         if items:
-            values = [dict(batch_id=self.id, ia_id=item) for item in items]
-            db.get_db().multiple_insert("import_item", values)
+            # Either create a reference to an IA id which will be loaded
+            # from Archive.org metadata, or provide json data book record
+            # which will be loaded directly into the OL catalog
+            values = [
+                {
+                    'batch_id': self.id,
+                    **({'ia_id': item} if ia_items else {'data': json.dumps(item)})
+                }
+                for item in items
+            ]
+            try:
+                # Upgrads psql and use `INSERT OR IGNORE`
```

Comment:
Is this comment still necessary?  Ther's the dedupe method above, and the exception handling also - which one solves the problem, or are both needed?

Reply:
1. The dedupe method is only implemented for the `ia_id` field (not `data`)
2. The exception handling works, but ideally we'd upgrade psql and use the `insert or ignore` syntax

Reply:
```suggestion
                # TODO: Upgrade psql and use `INSERT OR IGNORE`
```